<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Damiano Oldoni" />

<meta name="date" content="2022-01-25" />

<title>Darwin Core mapping</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="static/css/riparias.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">union-list</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="dwc_mapping.html">Darwin Core mapping</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/riparias/union-list">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Darwin Core mapping</h1>
<h3 class="subtitle">For: List of Invasive Alien Species of Union concern</h3>
<h4 class="author">Damiano Oldoni</h4>
<h4 class="date">2022-01-25</h4>

</div>


<div id="setup" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Setup</h1>
<p>Install required libraries (only if the libraries have not been installed before):</p>
<pre class="r"><code>installed &lt;- rownames(installed.packages())
required &lt;- c(&quot;tidyverse&quot;, &quot;tidylog&quot;, &quot;magrittr&quot;, &quot;here&quot;, &quot;janitor&quot;, &quot;digest&quot;, &quot;rgbif&quot;)
if (!all(required %in% installed)) {
  install.packages(required[!required %in% installed])
}</code></pre>
<p>Load libraries:</p>
<pre class="r"><code>library(tidyverse)      # To do data science
library(tidylog)        # To provide feedback on dplyr functions
library(magrittr)       # To use %&lt;&gt;% pipes
library(here)           # To find files
library(janitor)        # To clean input data
library(digest)         # To generate hashes
library(rgbif)          # To use GBIF services</code></pre>
</div>
<div id="read-source-data" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Read source data</h1>
<p>Create a data frame <code>input_data</code> from the source data:</p>
<pre class="r"><code>input_data &lt;- read_tsv(file = here(&quot;data&quot;, &quot;raw&quot;, &quot;list_of_invasive_alien_species_of_union_concern.tsv&quot;)) </code></pre>
<pre><code>## Rows: 66 Columns: 5</code></pre>
<pre><code>## -- Column specification --------------------------------------------------------
## Delimiter: &quot;\t&quot;
## chr (5): scientific_name, scientific_name_alternative, english_name, entry_i...</code></pre>
<pre><code>## 
## i Use `spec()` to retrieve the full column specification for this data.
## i Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Preview data:</p>
<pre class="r"><code>input_data %&gt;% head(n = 5)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["scientific_name"],"name":[1],"type":["chr"],"align":["left"]},{"label":["scientific_name_alternative"],"name":[2],"type":["chr"],"align":["left"]},{"label":["english_name"],"name":[3],"type":["chr"],"align":["left"]},{"label":["entry_into_force"],"name":[4],"type":["chr"],"align":["left"]},{"label":["kingdom"],"name":[5],"type":["chr"],"align":["left"]}],"data":[{"1":"Acacia saligna (Labill.) H.L.Wendl.","2":"Acacia cyanophylla Lindl.","3":"Golden wreath wattle","4":"15 August 2019","5":"Plantae"},{"1":"Acridotheres tristis Linnaeus, 1766","2":"NA","3":"Common myna","4":"15 August 2019","5":"Animalia"},{"1":"Ailanthus altissima (Mill.) Swingle","2":"NA","3":"Tree of heaven","4":"15 August 2019","5":"Plantae"},{"1":"Alopochen aegyptiacus Linnaeus, 1766","2":"NA","3":"Egyptian goose","4":"2 August 2017","5":"Animalia"},{"1":"Alternanthera philoxeroides (Mart.) Griseb.","2":"NA","3":"Alligator weed","4":"2 August 2017","5":"Plantae"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="process-source-data" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Process source data</h1>
<div id="tidy-data" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Tidy data</h2>
<p>The column <code>scientific_name_alternative</code> contains recent taxonomic changes. We remove this column by making new rows for it and copying the information related to English vernacular name (column <code>english_name</code>), entry into force date (column <code>entry_into_force</code>) and kingdom (column <code>kingdom</code>):</p>
<pre class="r"><code>input_data %&lt;&gt;% 
  pivot_longer(starts_with(&quot;scientific_name&quot;),
               names_to = &quot;col_name&quot;, # help column
               values_to = &quot;scientific_name&quot;) %&gt;%
  select(-col_name) %&gt;%
  relocate(scientific_name) %&gt;%
  filter(!is.na(scientific_name))</code></pre>
<pre><code>## pivot_longer: reorganized (scientific_name_alternative) into (col_name) [was 66x5, now 132x5]</code></pre>
<pre><code>## select: dropped one variable (col_name)</code></pre>
<pre><code>## relocate: columns reordered (scientific_name, english_name, entry_into_force, kingdom)</code></pre>
<pre><code>## filter: removed 57 rows (43%), 75 rows remaining</code></pre>
<p>This operation results in the following data.frame:</p>
<pre class="r"><code>input_data</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["scientific_name"],"name":[1],"type":["chr"],"align":["left"]},{"label":["english_name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["entry_into_force"],"name":[3],"type":["chr"],"align":["left"]},{"label":["kingdom"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"Acacia saligna (Labill.) H.L.Wendl.","2":"Golden wreath wattle","3":"15 August 2019","4":"Plantae"},{"1":"Acacia cyanophylla Lindl.","2":"Golden wreath wattle","3":"15 August 2019","4":"Plantae"},{"1":"Acridotheres tristis Linnaeus, 1766","2":"Common myna","3":"15 August 2019","4":"Animalia"},{"1":"Ailanthus altissima (Mill.) Swingle","2":"Tree of heaven","3":"15 August 2019","4":"Plantae"},{"1":"Alopochen aegyptiacus Linnaeus, 1766","2":"Egyptian goose","3":"2 August 2017","4":"Animalia"},{"1":"Alternanthera philoxeroides (Mart.) Griseb.","2":"Alligator weed","3":"2 August 2017","4":"Plantae"},{"1":"Andropogon virginicus L.","2":"Broomsedge bluestem","3":"15 August 2019","4":"Plantae"},{"1":"Arthurdendyus triangulatus (Dendy, 1894) Jones & Gerard (1999)","2":"New Zealand flatworm","3":"15 August 2019","4":"Animalia"},{"1":"Asclepias syriaca L.","2":"Common milkweed","3":"2 August 2017","4":"Plantae"},{"1":"Baccharis halimifolia L.","2":"Eastern baccharis","3":"3 August 2016","4":"Plantae"},{"1":"Cabomba caroliniana Gray","2":"Fanwort","3":"3 August 2016","4":"Plantae"},{"1":"Callosciurus erythraeus Pallas, 1779","2":"Pallas' squirrel","3":"3 August 2016","4":"Animalia"},{"1":"Cardiospermum grandiflorum Sw.","2":"Balloon vine","3":"15 August 2019","4":"Plantae"},{"1":"Cortaderia jubata (Lemoine ex Carrière) Stapf","2":"Purple pampas grass","3":"15 August 2019","4":"Plantae"},{"1":"Corvus splendens Viellot, 1817","2":"Indian house crow","3":"3 August 2016","4":"Animalia"},{"1":"Ehrharta calycina Sm.","2":"Perrenial veldt grass","3":"15 August 2019","4":"Plantae"},{"1":"Eichhornia crassipes (Martius) Solms","2":"Water hyacinth","3":"3 August 2016","4":"Plantae"},{"1":"Elodea nuttallii (Planch.) St. John","2":"Nuttall's waterweed","3":"2 August 2017","4":"Plantae"},{"1":"Eriocheir sinensis H. Milne Edwards, 1854","2":"Chinese mittencrab","3":"3 August 2016","4":"Animalia"},{"1":"Gunnera tinctoria (Molina) Mirbel","2":"Chilean rhubarb","3":"2 August 2017","4":"Plantae"},{"1":"Gymnocoronis spilanthoides (D.Don ex Hook. & Arn.) DC.","2":"Senegal tea plant","3":"15 August 2019","4":"Plantae"},{"1":"Heracleum mantegazzianum Sommier & Levier","2":"Giant hogweed","3":"2 August 2017","4":"Plantae"},{"1":"Heracleum persicum Fischer","2":"Persian hogweed","3":"3 August 2016","4":"Plantae"},{"1":"Heracleum sosnowskyi Mandenova","2":"Sosnowsky's hogweed","3":"3 August 2016","4":"Plantae"},{"1":"Herpestes javanicus É. Geoffroy Saint-Hilaire, 1818","2":"Small Asian mongoose","3":"3 August 2016","4":"Animalia"},{"1":"Humulus scandens (Lour.) Merr.","2":"Japanese hop","3":"15 August 2019","4":"Plantae"},{"1":"Hydrocotyle ranunculoides L. f.","2":"Floating pennywort","3":"3 August 2016","4":"Plantae"},{"1":"Impatiens glandulifera Royle","2":"Himalayan balsam","3":"2 August 2017","4":"Plantae"},{"1":"Lagarosiphon major (Ridley) Moss","2":"Curly waterweed","3":"3 August 2016","4":"Plantae"},{"1":"Lepomis gibbosus Linnaeus, 1758","2":"Pumpkinseed","3":"15 August 2019","4":"Animalia"},{"1":"Lespedeza cuneata (Dum.Cours.) G.Don","2":"Chinese bushclover","3":"15 August 2019","4":"Plantae"},{"1":"Lespedeza juncea var. sericea (Thunb.) Lace & Hauech","2":"Chinese bushclover","3":"15 August 2019","4":"Plantae"},{"1":"Lithobates (Rana) catesbeianus Shaw, 1802","2":"American bullfrog","3":"3 August 2016","4":"Animalia"},{"1":"Ludwigia grandiflora (Michx.) Greuter & Burdet","2":"Water-primrose","3":"3 August 2016","4":"Plantae"},{"1":"Ludwigia peploides (Kunth) P.H. Raven","2":"Floating primrose-willow","3":"3 August 2016","4":"Plantae"},{"1":"Lygodium japonicum (Thunb.) Sw.","2":"Vine-like fern","3":"15 August 2019","4":"Plantae"},{"1":"Lysichiton americanus Hultén and St. John","2":"American skunk cabbage","3":"3 August 2016","4":"Plantae"},{"1":"Microstegium vimineum (Trin.) A. Camus","2":"Japanese stiltgrass","3":"2 August 2017","4":"Plantae"},{"1":"Muntiacus reevesi Ogilby, 1839","2":"Muntjac deer","3":"3 August 2016","4":"Animalia"},{"1":"Myocastor coypus Molina, 1782","2":"Coypu","3":"3 August 2016","4":"Animalia"},{"1":"Myriophyllum aquaticum (Vell.) Verdc.","2":"Parrot's feather","3":"3 August 2016","4":"Plantae"},{"1":"Myriophyllum heterophyllum Michaux","2":"Broadleaf watermilfoil","3":"2 August 2017","4":"Plantae"},{"1":"Nasua nasua Linnaeus, 1766","2":"Coati","3":"3 August 2016","4":"Animalia"},{"1":"Nyctereutes procyonoides Gray, 1834","2":"Raccoon dog","3":"2 February 2019","4":"Animalia"},{"1":"Ondatra zibethicus Linnaeus, 1766","2":"Muskrat","3":"2 August 2017","4":"Animalia"},{"1":"Orconectes limosus Rafinesque, 1817","2":"Spiny-cheek crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Faxonius limosus Rafinesque, 1817","2":"Spiny-cheek crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Orconectes virilis Hagen, 1870","2":"Virile crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Faxonius virilis Hagen, 1870","2":"Virile crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Oxyura jamaicensis Gmelin, 1789","2":"Ruddy duck","3":"3 August 2016","4":"Animalia"},{"1":"Pacifastacus leniusculus Dana, 1852","2":"Signal crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Parthenium hysterophorus L.","2":"Whitetop weed","3":"3 August 2016","4":"Plantae"},{"1":"Pennisetum setaceum (Forssk.) Chiov.","2":"Crimson fountaingrass","3":"2 August 2017","4":"Plantae"},{"1":"Perccottus glenii Dybowski, 1877","2":"Amur sleeper","3":"3 August 2016","4":"Animalia"},{"1":"Persicaria perfoliata (L.) H. Gross","2":"Asiatic tearthumb","3":"3 August 2016","4":"Plantae"},{"1":"Polygonum perfoliatum L.","2":"Asiatic tearthumb","3":"3 August 2016","4":"Plantae"},{"1":"Plotosus lineatus (Thunberg, 1787)","2":"Striped eel catfish","3":"15 August 2019","4":"Animalia"},{"1":"Procambarus clarkii Girard, 1852","2":"Red swamp crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Procambarus fallax (Hagen, 1870) f. virginalis","2":"Marbled crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Procambarus virginalis Lyko, 2017","2":"Marbled crayfish","3":"3 August 2016","4":"Animalia"},{"1":"Procyon lotor Linnaeus, 1758","2":"Raccoon","3":"3 August 2016","4":"Animalia"},{"1":"Prosopis juliflora (Sw.) DC.","2":"Mesquite","3":"15 August 2019","4":"Plantae"},{"1":"Pseudorasbora parva Temminck & Schlegel, 1846","2":"Stone moroko","3":"3 August 2016","4":"Animalia"},{"1":"Pueraria montana (Lour.) Merr. var. lobata (Willd.)","2":"Kudzu vine","3":"3 August 2016","4":"Plantae"},{"1":"Pueraria lobata (Willd.) Ohwi","2":"Kudzu vine","3":"3 August 2016","4":"Plantae"},{"1":"Salvinia molesta D.S. Mitch.","2":"Salvinia moss","3":"15 August 2019","4":"Plantae"},{"1":"Salvinia adnata Desv.","2":"Salvinia moss","3":"15 August 2019","4":"Plantae"},{"1":"Sciurus carolinensis Gmelin,1788","2":"Grey squirrel","3":"3 August 2016","4":"Animalia"},{"1":"Sciurus niger Linnaeus, 1758","2":"Fox squirrel","3":"3 August 2016","4":"Animalia"},{"1":"Tamias sibiricus Laxmann, 1769","2":"Siberian chipmunk","3":"3 August 2016","4":"Animalia"},{"1":"Threskiornis aethiopicus Latham, 1790","2":"Sacred ibis","3":"3 August 2016","4":"Animalia"},{"1":"Trachemys scripta Schoepff, 1792","2":"Red-eared, yellow-bellied and Cumberland sliders","3":"3 August 2016","4":"Animalia"},{"1":"Triadica sebifera (L.) Small","2":"Chinese tallow","3":"15 August 2019","4":"Plantae"},{"1":"Sapium sebiferum (L.) Roxb.","2":"Chinese tallow","3":"15 August 2019","4":"Plantae"},{"1":"Vespa velutina nigrithorax de Buysson, 1905","2":"Asian hornet","3":"3 August 2016","4":"Animalia"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="scientific-names" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Scientific names</h2>
<p>Use the <a href="https://www.gbif.org/tools/name-parser">GBIF nameparser</a> to retrieve nomenclatural information for the scientific names in the checklist:</p>
<pre class="r"><code>parsed_names &lt;- input_data %&gt;%
  distinct(scientific_name) %&gt;%
  pull() %&gt;% # Create vector from dataframe
  parsenames() # An rgbif function</code></pre>
<pre><code>## distinct: no rows removed</code></pre>
<p>Show scientific names with nomenclatural issues, i.e. not of <code>type = SCIENTIFIC</code> or that could not be fully parsed. Note: these are not necessarily incorrect.</p>
<pre class="r"><code>parsed_names %&gt;%
  select(scientificname, type, parsed, parsedpartially, rankmarker) %&gt;%
  filter(!(type == &quot;SCIENTIFIC&quot; &amp; parsed == &quot;TRUE&quot; &amp; parsedpartially == &quot;FALSE&quot;))</code></pre>
<pre><code>## select: dropped 11 variables (genusorabove, specificepithet, authorship, bracketauthorship, canonicalname, …)</code></pre>
<pre><code>## filter: removed all rows (100%)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["scientificname"],"name":[1],"type":["chr"],"align":["left"]},{"label":["type"],"name":[2],"type":["chr"],"align":["left"]},{"label":["parsed"],"name":[3],"type":["lgl"],"align":["right"]},{"label":["parsedpartially"],"name":[4],"type":["lgl"],"align":["right"]},{"label":["rankmarker"],"name":[5],"type":["chr"],"align":["left"]}],"data":[],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="taxon-ranks" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Taxon ranks</h2>
<p>The nameparser function also provides information about the rank of the taxon (in <code>rankmarker</code>). Here we join this information with our checklist. Cleaning these ranks will be done in the Taxon Core mapping:</p>
<pre class="r"><code>input_data %&lt;&gt;% left_join(
  parsed_names %&gt;%
  select(scientificname, rankmarker),
  by = c(&quot;scientific_name&quot; = &quot;scientificname&quot;))</code></pre>
<pre><code>## select: dropped 14 variables (type, genusorabove, specificepithet, authorship, bracketauthorship, …)</code></pre>
<pre><code>## left_join: added one column (rankmarker)</code></pre>
<pre><code>##            &gt; rows only in x    0</code></pre>
<pre><code>##            &gt; rows only in y  ( 0)</code></pre>
<pre><code>##            &gt; matched rows     75</code></pre>
<pre><code>##            &gt;                 ====</code></pre>
<pre><code>##            &gt; rows total       75</code></pre>
</div>
<div id="taxon-ids" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Taxon IDs</h2>
<p>To link taxa with information in the extension(s), each taxon needs a unique and relatively stable <code>taxonID</code>. Here we create one in the form of <code>dataset_shortname:taxon:hash</code>, where <code>hash</code> is unique code based on scientific name and kingdom (that will remain the same as long as scientific name and kingdom remain the same):</p>
<pre class="r"><code>vdigest &lt;- Vectorize(digest) # Vectorize digest function to work with vectors
input_data %&lt;&gt;% mutate(taxon_id = paste(
  &quot;union-list&quot;,
  &quot;taxon&quot;,
  vdigest(paste(scientific_name, kingdom), algo = &quot;md5&quot;),
  sep = &quot;:&quot;
))</code></pre>
<pre><code>## mutate: new variable &#39;taxon_id&#39; (character) with 75 unique values and 0% NA</code></pre>
</div>
<div id="preview-data" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Preview data</h2>
<p>Show the number of taxa and distributions per kingdom and rank:</p>
<pre class="r"><code>input_data %&gt;%
  group_by(kingdom, rankmarker) %&gt;%
  summarize(
    `# taxa` = n_distinct(taxon_id),
    `# distributions` = n()
  ) %&gt;%
  adorn_totals(&quot;row&quot;)</code></pre>
<pre><code>## group_by: 2 grouping variables (kingdom, rankmarker)</code></pre>
<pre><code>## summarize: now 5 rows and 4 columns, one group variable remaining (kingdom)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["kingdom"],"name":[1],"type":["chr"],"align":["left"]},{"label":["rankmarker"],"name":[2],"type":["chr"],"align":["left"]},{"label":["# taxa"],"name":[3],"type":["int"],"align":["right"]},{"label":["# distributions"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"Animalia","2":"f.","3":"1","4":"1"},{"1":"Animalia","2":"infrasp.","3":"1","4":"1"},{"1":"Animalia","2":"sp.","3":"31","4":"31"},{"1":"Plantae","2":"sp.","3":"40","4":"40"},{"1":"Plantae","2":"var.","3":"2","4":"2"},{"1":"Total","2":"-","3":"75","4":"75"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Preview data:</p>
<pre class="r"><code>input_data %&gt;% head()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["scientific_name"],"name":[1],"type":["chr"],"align":["left"]},{"label":["english_name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["entry_into_force"],"name":[3],"type":["chr"],"align":["left"]},{"label":["kingdom"],"name":[4],"type":["chr"],"align":["left"]},{"label":["rankmarker"],"name":[5],"type":["chr"],"align":["left"]},{"label":["taxon_id"],"name":[6],"type":["chr"],"align":["left"]}],"data":[{"1":"Acacia saligna (Labill.) H.L.Wendl.","2":"Golden wreath wattle","3":"15 August 2019","4":"Plantae","5":"sp.","6":"union-list:taxon:e75968a892a25181e92a7a16e2480191"},{"1":"Acacia cyanophylla Lindl.","2":"Golden wreath wattle","3":"15 August 2019","4":"Plantae","5":"sp.","6":"union-list:taxon:49a114c68be9999ff8cb203bc418d9df"},{"1":"Acridotheres tristis Linnaeus, 1766","2":"Common myna","3":"15 August 2019","4":"Animalia","5":"sp.","6":"union-list:taxon:88fab226290e72ddf6e538d934aa58a6"},{"1":"Ailanthus altissima (Mill.) Swingle","2":"Tree of heaven","3":"15 August 2019","4":"Plantae","5":"sp.","6":"union-list:taxon:dc1a9c7d8ecbea2c17659b65aac590ef"},{"1":"Alopochen aegyptiacus Linnaeus, 1766","2":"Egyptian goose","3":"2 August 2017","4":"Animalia","5":"sp.","6":"union-list:taxon:52f42b74b3d05c74023d8a75c402989d"},{"1":"Alternanthera philoxeroides (Mart.) Griseb.","2":"Alligator weed","3":"2 August 2017","4":"Plantae","5":"sp.","6":"union-list:taxon:b7674f5e6e0f6beed1f1360ed992f885"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div id="taxon-core" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Taxon core</h1>
<p>Create a <code>taxon</code> dataframe starting from <code>input_data</code>:</p>
<pre class="r"><code>taxon &lt;- input_data</code></pre>
<div id="term-mapping" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Term mapping</h2>
<p>Map the data to <a href="http://rs.gbif.org/core/dwc_taxon_2015-04-24.xml">Darwin Core Taxon</a>.</p>
<p>Start with record-level terms which contain metadata about the dataset (which is generally the same for all records).</p>
<div id="language" class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> language</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_language = &quot;en&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_language&#39; (character) with one unique value and 0% NA</code></pre>
</div>
<div id="license" class="section level3" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> license</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_license = &quot;http://creativecommons.org/publicdomain/zero/1.0/&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_license&#39; (character) with one unique value and 0% NA</code></pre>
</div>
<div id="rightsholder" class="section level3" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> rightsHolder</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_rightsHolder = &quot;INBO&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_rightsHolder&#39; (character) with one unique value and 0% NA</code></pre>
</div>
<div id="datasetid" class="section level3" number="4.1.4">
<h3><span class="header-section-number">4.1.4</span> datasetID</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_datasetID = &quot;&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_datasetID&#39; (character) with one unique value and 0% NA</code></pre>
</div>
<div id="institutioncode" class="section level3" number="4.1.5">
<h3><span class="header-section-number">4.1.5</span> institutionCode</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_institutionCode = &quot;INBO&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_institutionCode&#39; (character) with one unique value and 0% NA</code></pre>
</div>
<div id="datasetname" class="section level3" number="4.1.6">
<h3><span class="header-section-number">4.1.6</span> datasetName</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_datasetName = &quot;List of Invasive Alien Species of Union concern&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_datasetName&#39; (character) with one unique value and 0% NA</code></pre>
<p>The following terms contain information about the taxon:</p>
</div>
<div id="taxonid" class="section level3" number="4.1.7">
<h3><span class="header-section-number">4.1.7</span> taxonID</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_taxonID = taxon_id)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_taxonID&#39; (character) with 75 unique values and 0% NA</code></pre>
</div>
<div id="scientificname" class="section level3" number="4.1.8">
<h3><span class="header-section-number">4.1.8</span> scientificName</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_scientificName = scientific_name)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_scientificName&#39; (character) with 75 unique values and 0% NA</code></pre>
</div>
<div id="kingdom" class="section level3" number="4.1.9">
<h3><span class="header-section-number">4.1.9</span> kingdom</h3>
<p>Inspect values:</p>
<pre class="r"><code>taxon %&gt;%
  group_by(kingdom) %&gt;%
  count()</code></pre>
<pre><code>## group_by: one grouping variable (kingdom)</code></pre>
<pre><code>## count: now 2 rows and 2 columns, one group variable remaining (kingdom)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["kingdom"],"name":[1],"type":["chr"],"align":["left"]},{"label":["n"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"Animalia","2":"33"},{"1":"Plantae","2":"42"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Map values:</p>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_kingdom = kingdom)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_kingdom&#39; (character) with 2 unique values and 0% NA</code></pre>
</div>
<div id="taxonrank" class="section level3" number="4.1.10">
<h3><span class="header-section-number">4.1.10</span> taxonRank</h3>
<p>Inspect values:</p>
<pre class="r"><code>taxon %&gt;%
  group_by(rankmarker) %&gt;%
  count()</code></pre>
<pre><code>## group_by: one grouping variable (rankmarker)</code></pre>
<pre><code>## count: now 4 rows and 2 columns, one group variable remaining (rankmarker)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["rankmarker"],"name":[1],"type":["chr"],"align":["left"]},{"label":["n"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"f.","2":"1"},{"1":"infrasp.","2":"1"},{"1":"sp.","2":"71"},{"1":"var.","2":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Map values by recoding to the <a href="http://rs.gbif.org/vocabulary/gbif/rank_2015-04-24.xml">GBIF rank vocabulary</a>:</p>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_taxonRank = recode(rankmarker,
  &quot;agg.&quot;      = &quot;speciesAggregate&quot;,
  &quot;f.&quot;         = &quot;form&quot;,
  &quot;infrasp.&quot;  = &quot;infraspecificname&quot;,
  &quot;sp.&quot;       = &quot;species&quot;,
  &quot;var.&quot;      = &quot;variety&quot;,
  .default    = &quot;&quot;,
  .missing    = &quot;&quot;
))</code></pre>
<pre><code>## mutate: new variable &#39;dwc_taxonRank&#39; (character) with 4 unique values and 0% NA</code></pre>
<p>Inspect mapped values:</p>
<pre class="r"><code>taxon %&gt;%
  group_by(rankmarker, dwc_taxonRank) %&gt;%
  count()</code></pre>
<pre><code>## group_by: 2 grouping variables (rankmarker, dwc_taxonRank)</code></pre>
<pre><code>## count: now 4 rows and 3 columns, 2 group variables remaining (rankmarker, dwc_taxonRank)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["rankmarker"],"name":[1],"type":["chr"],"align":["left"]},{"label":["dwc_taxonRank"],"name":[2],"type":["chr"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"f.","2":"form","3":"1"},{"1":"infrasp.","2":"infraspecificname","3":"1"},{"1":"sp.","2":"species","3":"71"},{"1":"var.","2":"variety","3":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="nomenclaturalcode" class="section level3" number="4.1.11">
<h3><span class="header-section-number">4.1.11</span> nomenclaturalCode</h3>
<pre class="r"><code>taxon %&lt;&gt;% mutate(dwc_nomenclaturalCode = &quot;ICZN&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_nomenclaturalCode&#39; (character) with one unique value and 0% NA</code></pre>
</div>
</div>
<div id="post-processing" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Post-processing</h2>
<p>Only keep the Darwin Core columns:</p>
<pre class="r"><code>taxon %&lt;&gt;% select(starts_with(&quot;dwc_&quot;))</code></pre>
<pre><code>## select: dropped 6 variables (scientific_name, english_name, entry_into_force, kingdom, rankmarker, …)</code></pre>
<p>Drop the <code>dwc_</code> prefix:</p>
<pre class="r"><code>colnames(taxon) &lt;- str_remove(colnames(taxon), &quot;dwc_&quot;)</code></pre>
<p>Preview data:</p>
<pre class="r"><code>taxon %&gt;% head()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["language"],"name":[1],"type":["chr"],"align":["left"]},{"label":["license"],"name":[2],"type":["chr"],"align":["left"]},{"label":["rightsHolder"],"name":[3],"type":["chr"],"align":["left"]},{"label":["datasetID"],"name":[4],"type":["chr"],"align":["left"]},{"label":["institutionCode"],"name":[5],"type":["chr"],"align":["left"]},{"label":["datasetName"],"name":[6],"type":["chr"],"align":["left"]},{"label":["taxonID"],"name":[7],"type":["chr"],"align":["left"]},{"label":["scientificName"],"name":[8],"type":["chr"],"align":["left"]},{"label":["kingdom"],"name":[9],"type":["chr"],"align":["left"]},{"label":["taxonRank"],"name":[10],"type":["chr"],"align":["left"]},{"label":["nomenclaturalCode"],"name":[11],"type":["chr"],"align":["left"]}],"data":[{"1":"en","2":"http://creativecommons.org/publicdomain/zero/1.0/","3":"INBO","4":"","5":"INBO","6":"List of Invasive Alien Species of Union concern","7":"union-list:taxon:e75968a892a25181e92a7a16e2480191","8":"Acacia saligna (Labill.) H.L.Wendl.","9":"Plantae","10":"species","11":"ICZN"},{"1":"en","2":"http://creativecommons.org/publicdomain/zero/1.0/","3":"INBO","4":"","5":"INBO","6":"List of Invasive Alien Species of Union concern","7":"union-list:taxon:49a114c68be9999ff8cb203bc418d9df","8":"Acacia cyanophylla Lindl.","9":"Plantae","10":"species","11":"ICZN"},{"1":"en","2":"http://creativecommons.org/publicdomain/zero/1.0/","3":"INBO","4":"","5":"INBO","6":"List of Invasive Alien Species of Union concern","7":"union-list:taxon:88fab226290e72ddf6e538d934aa58a6","8":"Acridotheres tristis Linnaeus, 1766","9":"Animalia","10":"species","11":"ICZN"},{"1":"en","2":"http://creativecommons.org/publicdomain/zero/1.0/","3":"INBO","4":"","5":"INBO","6":"List of Invasive Alien Species of Union concern","7":"union-list:taxon:dc1a9c7d8ecbea2c17659b65aac590ef","8":"Ailanthus altissima (Mill.) Swingle","9":"Plantae","10":"species","11":"ICZN"},{"1":"en","2":"http://creativecommons.org/publicdomain/zero/1.0/","3":"INBO","4":"","5":"INBO","6":"List of Invasive Alien Species of Union concern","7":"union-list:taxon:52f42b74b3d05c74023d8a75c402989d","8":"Alopochen aegyptiacus Linnaeus, 1766","9":"Animalia","10":"species","11":"ICZN"},{"1":"en","2":"http://creativecommons.org/publicdomain/zero/1.0/","3":"INBO","4":"","5":"INBO","6":"List of Invasive Alien Species of Union concern","7":"union-list:taxon:b7674f5e6e0f6beed1f1360ed992f885","8":"Alternanthera philoxeroides (Mart.) Griseb.","9":"Plantae","10":"species","11":"ICZN"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Save to CSV:</p>
<pre class="r"><code>write_csv(taxon, here(&quot;data&quot;, &quot;processed&quot;, &quot;taxon.csv&quot;), na = &quot;&quot;)</code></pre>
</div>
</div>
<div id="map-vernacular-names-extension" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Map vernacular names extension</h1>
<div id="term-mapping-1" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Term mapping</h2>
<p>Create a <code>vernacular</code> dataframe starting from <code>input_data</code>:</p>
<pre class="r"><code>vernacular &lt;- input_data</code></pre>
<div id="taxonid-1" class="section level3" number="5.1.1">
<h3><span class="header-section-number">5.1.1</span> TaxonID</h3>
<pre class="r"><code>vernacular %&lt;&gt;% mutate(dwc_taxonID = taxon_id)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_taxonID&#39; (character) with 75 unique values and 0% NA</code></pre>
</div>
<div id="vernacularname" class="section level3" number="5.1.2">
<h3><span class="header-section-number">5.1.2</span> vernacularName</h3>
<pre class="r"><code>vernacular %&lt;&gt;% mutate(dwc_vernacularName = english_name)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_vernacularName&#39; (character) with 66 unique values and 0% NA</code></pre>
</div>
<div id="source" class="section level3" number="5.1.3">
<h3><span class="header-section-number">5.1.3</span> source</h3>
<p>The <code>source</code> is based on expert knowledge, set to <code>NA</code>:</p>
<pre class="r"><code>vernacular %&lt;&gt;% mutate(dwc_source = NA)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_source&#39; (logical) with one unique value and 100% NA</code></pre>
</div>
<div id="language-1" class="section level3" number="5.1.4">
<h3><span class="header-section-number">5.1.4</span> language</h3>
<p>Set language code to <code>en</code> (English):</p>
<pre class="r"><code>vernacular %&lt;&gt;% mutate(dwc_language = &quot;en&quot;)</code></pre>
<pre><code>## mutate: new variable &#39;dwc_language&#39; (character) with one unique value and 0% NA</code></pre>
</div>
</div>
<div id="post-processing-1" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Post-processing</h2>
<p>Only keep the Darwin Core columns:</p>
<pre class="r"><code>vernacular %&lt;&gt;% select(starts_with(&quot;dwc_&quot;))</code></pre>
<pre><code>## select: dropped 6 variables (scientific_name, english_name, entry_into_force, kingdom, rankmarker, …)</code></pre>
<p>Drop the <code>dwc_</code> prefix:</p>
<pre class="r"><code>colnames(vernacular) &lt;- str_remove(colnames(vernacular), &quot;dwc_&quot;)</code></pre>
<p>Preview data:</p>
<pre class="r"><code>vernacular %&gt;% head()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["taxonID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["vernacularName"],"name":[2],"type":["chr"],"align":["left"]},{"label":["source"],"name":[3],"type":["lgl"],"align":["right"]},{"label":["language"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"union-list:taxon:e75968a892a25181e92a7a16e2480191","2":"Golden wreath wattle","3":"NA","4":"en"},{"1":"union-list:taxon:49a114c68be9999ff8cb203bc418d9df","2":"Golden wreath wattle","3":"NA","4":"en"},{"1":"union-list:taxon:88fab226290e72ddf6e538d934aa58a6","2":"Common myna","3":"NA","4":"en"},{"1":"union-list:taxon:dc1a9c7d8ecbea2c17659b65aac590ef","2":"Tree of heaven","3":"NA","4":"en"},{"1":"union-list:taxon:52f42b74b3d05c74023d8a75c402989d","2":"Egyptian goose","3":"NA","4":"en"},{"1":"union-list:taxon:b7674f5e6e0f6beed1f1360ed992f885","2":"Alligator weed","3":"NA","4":"en"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Save to CSV:</p>
<pre class="r"><code>write_csv(vernacular, here(&quot;data&quot;, &quot;processed&quot;, &quot;vernacularname.csv&quot;), na = &quot;&quot;)</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
